### PR TITLE
Use dogpile.cache + pylru instead of functools.lru_cache, rewrite YumFinder

### PR DESCRIPTION
**Pull request by 0xDEC0DE**
Pull in `dogpile.cache` and allow SouFi to be initialized with any provided, custom, etc. backends that may be available.

Also rewrite a fair amount of the `YumFinder` to make it compatible with all of the backends provided by `dogpile.cache`.

Fixes: Issue #18

### CODE REVIEW

**Review by cburgess — APPROVED** [review]
This is the way.

File: soufi/finder.py [hunk 1]
```diff
@@ -172,12 +178,27 @@ def distro(self) -> Union[Distro, str]:
...
+        cache_backend: str = None,
+        cache_args: Iterable[str] = None,
     ):
         self.name = name
         self.version = version
         self.s_type = s_type
 
+        # By default, use a small in-memory LRU cache, 512 objects large
+        if not cache_backend:
+            cache_backend = 'dogpile.cache.memory'
```
Comment (juledwar) [inline soufi/finder.py:194]: This should be null by default, not memory.

Comment (0xDEC0DE) [reply soufi/finder.py:194]: That's fair; I was aiming for a drop-in replacement for `functools.lru_cache`, but this would make the implementation simpler.

File: soufi/finder.py [hunk 2]
```diff
@@ -172,12 +178,27 @@ def distro(self) -> Union[Distro, str]:
...
+            if not cache_args:
+                cache_args = dict(cache_dict=pylru.lrucache(size=512))
```
Comment (juledwar) [inline soufi/finder.py:196]: How big is an "object"?

**Review by juledwar — COMMENTED** [review]
We also need to consider cache invalidation.

File: soufi/finder.py [hunk 3]
```diff
@@ -161,6 +163,10 @@ class SourceFinder(metaclass=abc.ABCMeta):
     :param name: Name of the source package to find
     :param version: Version of the source package to find
     :param s_type: A SourceType indicating the type of package
+    :param cache_backend: A string containing the name of a registered
+        dogpile.cache backend.
```
Comment (juledwar) [inline soufi/finder.py:167]: This needs to document the default if not specified.

File: soufi/finder.py (hunk 1 above)
Comment (juledwar) [reply soufi/finder.py:194]: I would also say, we could consider passing `dogpile.cache.memory` from the CLI to keep the current behaviour on there.

Comment (0xDEC0DE) [reply soufi/finder.py:194]: Absolutely.

File: soufi/finders/yum.py [hunk 4]
```diff
@@ -137,13 +136,13 @@ def _nevra_or_none(self, package):
...
-    @functools.lru_cache(maxsize=1024)
-    def _get_repo(cls, url):
+    def _get_repo(self, url):
         if not url.endswith('/'):
             url += '/'
         try:
-            return repomd.load(url)
+            return self._cache.get_or_create(
+                f"repo-{url}", repomd.load, creator_args=([url], {})
+            )
```
Comment (0xDEC0DE) [inline soufi/finders/yum.py:145]: How unfortunate: these objects are effectively `ElementTree` objects, which very much do not like to be serialized, so this will not work with e.g., `dogpile.cache.memorypickle` or `dogpile.cache.redis`.

The `repomd` library went with simplicity over configurability, so we'll have to write our own `load` method if we want to cache the inputs instead.  Either way, this should not merge as-is.

File: soufi/finders/yum.py [hunk 5]
```diff
@@ -224,3 +220,69 @@ def populate_archive(self, *args, **kwargs):  # pragma: no cover
...
+#  storage utilization low, and in the case of Redis caching, to reduce the
+#  amount of cache traffic sent over the wire.
+def get_repomd(queue, url):
+    try:
+        repo = repomd.load(url)
+    except urllib.error.HTTPError:
+        queue.put((None, None))
+        return
+    payload = repomd.defusedxml.lxml.tostring(repo._metadata)
+    queue.put((str(repo.baseurl), lzma.compress(payload)))
```
Comment (0xDEC0DE) [inline soufi/finders/yum.py:261]: This is missing the URL cleanup from the previous iteration, and the functests fall over because of it.  Will fix.

**Review by juledwar — COMMENTED** [review]
I'll review more later, probably on Monday, but it looks great so far. What a slog it's been.

File: soufi/finders/alpine.py [hunk 6]
```diff
@@ -66,20 +65,25 @@ def __init__(self, *args, aports_dir, **kwargs):
...
+            apkbuilds = {}
+            for apkbuild in glob.iglob(
+                f"{aports_dir}/**/*/APKBUILD", recursive=True
+            ):
+                path = Path(apkbuild)
+                package = path.parent.name
+                apkbuilds[package] = path
+            return apkbuilds
+
+        return self._cache.get_or_create(
```
[the remark this answers is not in the document]
Comment (0xDEC0DE) [reply soufi/finders/alpine.py:84]: Alpine trees tend to sprawl, and populating this list does take a while; I suspect that even in the worst-case it's a wash.

File: soufi/finder.py [hunk 7]
```diff
@@ -218,6 +222,46 @@ def find(
...
+            response = requests.get(
+                url, params=params, timeout=self.TIMEOUT, **kwargs
+            )
+            if response.status_code != requests.codes.ok:
+                raise exceptions.DownloadError(response.reason)
+            return response
+
+        key = url
+        if params:
+            key += f"?{params}"
```
Comment (juledwar) [inline soufi/finder.py:260]: Does this and `_head_url` intentionally ignore kwargs?

Comment (0xDEC0DE) [reply soufi/finder.py:260]: They don't, and they won't, because `kwargs` will be in-scope for the closure.  But it is confusing as-written; I don't know why I didn't go all-in on using the closure and have it not accept any inner args at all.

Will fix.

File: soufi/finders/ubuntu.py [hunk 8]
```diff
@@ -70,7 +68,7 @@ def populate_archive(self, temp_dir, tar):
         names = [url.rsplit('/', 1)[-1] for url in self.urls]
         for name, url in zip(names, self.urls):
             arcfile_name = self.download_file(
-                temp_dir, name, url, timeout=API_TIMEOUT
+                temp_dir, name, url, timeout=UbuntuFinder.TIMEOUT
```
Comment (juledwar) [inline soufi/finders/ubuntu.py:71]: Ok so the changes in this file have solidified a spark of a thought I was having from my earlier comments about the TIMEOUT value.
Instead of using a class property in the base class, let's make it an instance property that's set in `__init__` with a default of 30, and then just refer to self.timeout everywhere.

**Review by juledwar — COMMENTED** [review]
Moar feedback. Still need to look at tests.

File: soufi/finder.py (hunk 7 above)
Comment (juledwar) [reply soufi/finder.py:258]: Yeah I was about to say why use a closure at all, but this works too.

File: soufi/finders/alpine.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (0xDEC0DE) [reply soufi/finders/alpine.py]: On my workstation:
```
DEBUG:dogpile.cache.region:No value present for key: 'glob-/var/folders/w_/dl6trrf14556hcmt3rbl__640000gn/T/tmp66ut9t9j.aports'
DEBUG:dogpile.cache.region:Cache value generated in 1.111 seconds for key(s): 'glob-/var/folders/w_/dl6trrf14556hcmt3rbl__640000gn/T/tmp66ut9t9j.aports'
```

Cloning the `aports` directory is the part that takes up all the time, but saving ~1 second per source adds up on bulk operations.

File: soufi/finders/ubuntu.py (hunk 8 above)
Comment (0xDEC0DE) [reply soufi/finders/ubuntu.py:71]: I considered it, and it means passing the value from `SourceFinder`-derived objects into `DiscoverdSource`-derived objects, which means fidgeting with method signatures everywhere.

I figured class properties would look cleaner, but looking back at it, it's all pretty gnarly.  An instance property would also facilitate overriding it, which might be handy.

Comment (juledwar) [reply soufi/finders/ubuntu.py:71]: Yeah overriding was my main thought. It should be easy enough to pass the same value into the `DiscoveredSource`.

File: soufi/finders/alpine.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (juledwar) [reply soufi/finders/alpine.py]: Ok cool, that's worth it I guess.

File: soufi/finders/java.py [hunk 9]
```diff
@@ -22,21 +21,30 @@ def _find(self):
         source_url = self.get_source_url()
         return JavaDiscoveredSource([source_url])
 
+    # NOTE(nic): override this method to also return the response object so
+    #  that it may be inspected.
```
[the remark this answers is not in the document]
Comment (0xDEC0DE) [reply soufi/finders/java.py:25]: It's a one-off, and not particularly useful/interesting to any other call sites, so not worth passing around elsewhere.

Making a multi-value return the default behavior just feels like a great way to generate `ValueError`s later.  And my preferred option (a bespoke object that evaluates truthiness normally, but also possesses properties) seems a little bit bonkers here.

[the remark this answers is not in the document]
Comment (juledwar) [reply soufi/finders/java.py:25]: There's already a bespoke object — the `Response` ! Overriding here doesn't make any sense to me when you can do any of:
a) just return the Response everywhere.
b) raise an exception in the case the url is not found
c) return None in the case the URL is not found and always return the URL

I think it's best to just return the `Response` and raise an exception if it's not found. If you want the least work, return the Response or a None instead of True/False.

Sorry for nitpicking but this is going to be part of the public API on the base finder so it needs to be spot on.

**Review by juledwar — COMMENTED** [review]
Niiiiice. I just want to push back more on that overridden method, I really think you don't need it.

File: soufi/finders/java.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (0xDEC0DE) [reply soufi/finders/java.py:25]: nathan-fillion-meme

I don't really know what I was thinking before.

File: soufi/tests/finders/test_ubuntu_finder.py [hunk 10]
```diff
@@ -18,25 +18,13 @@
...
         return ubuntu.UbuntuFinder(name, version, SourceType.os)
 
-    def test_init_auths_to_Launchpad_only_once(self):
-        login = self.patch(ubuntu.Launchpad, 'login_anonymously')
-        self.make_finder()
-        self.make_finder()
-        login.assert_called_once_with(
-            "soufi", "production", mock.ANY, version="devel", timeout=30
-        )
-
```
Comment (juledwar) [inline soufi/tests/finders/test_ubuntu_finder.py:39]: Why did you remove this?

File: soufi/tests/finders/test_yum_finder.py [hunk 11]
```diff
@@ -338,7 +321,83 @@ def test__test_url_timeout(self):
...
+
+    def test_do_task(self):
+        data = self.factory.make_string('response')
+        queue = self.patch(yum, 'Queue')
+        queue.return_value.get.return_value = data
+        process = self.patch(yum, 'Process')
+        process.return_value.is_alive.return_value = True
+        response = yum.do_task('a', 'b', 'c')
+        self.assertEqual(data, response)
+        process.return_value.terminate.assert_called_once_with()
```
Comment (juledwar) [inline soufi/tests/finders/test_yum_finder.py:397]: I think these tests would read better and look more focused if you refactor the mock set up into a function that returns them. It's quite hard to see what's being tested; the inputs, outputs and assertions need to be very clear. Perhaps add some comments if the long setup is unavoidable so test intent is more obvious. For example, this `test_do_task` doesn't tell me *what* it's testing, just that it's calling it.

File: soufi/tests/functional/test_functional.py [hunk 12]
```diff
@@ -1,86 +1,266 @@
...
+
+# These tests will use the `memory_pickle` cache backend, which
+# serializes the cache payload before storing it; this should surface
+# any obvious problems with the other serializing cache backends
+# without forcing them as a dependency.
+
 
 class FunctionalFinderTests(testtools.TestCase):
     def setUp(self):
         self.skipTest("functional tests disabled for unit test runs")
```
Comment (juledwar) [inline soufi/tests/functional/test_functional.py:25]: Maybe add a tox `[functional]` target to run these.

**Review by juledwar — CHANGES_REQUESTED** [review]
This is looking really good! Thanks especially for adding all the extra functional tests for the other finders.  I have questions and suggestions about tests now (you expected that eventually, right? :laughing: )

File: soufi/tests/finders/test_ubuntu_finder.py (hunk 10 above)
Comment (0xDEC0DE) [reply soufi/tests/finders/test_ubuntu_finder.py:39]: Because unit tests run with the `dogpile.cache.null` backend, so `login_anonymously` will be called twice.

I could update the test to ensure that in this setup, `login_anonymously` is called twice, but that's not really adding much value.

File: soufi/tests/functional/test_functional.py (hunk 12 above)
Comment (0xDEC0DE) [reply soufi/tests/functional/test_functional.py:25]: Cool.  I was deferring any real action until I knew if you wanted these run with the main tests or not.  I figured you'd have opinions.  😆

File: soufi/tests/finders/test_yum_finder.py (hunk 11 above)
Comment (0xDEC0DE) [reply soufi/tests/finders/test_yum_finder.py:397]: I suppose I did get kind of spartan with explaining my tests near the end of this.  I'll see what I can do.

File: soufi/tests/finders/test_ubuntu_finder.py (hunk 10 above)
Comment (juledwar) [reply soufi/tests/finders/test_ubuntu_finder.py:39]: The problem is that it's removed a valid test. Can you set the cache to memory just for that call? It's a *very* slow API call on LP so making sure it's been cached is very valid!

File: soufi/tests/functional/test_functional.py [hunk 13]
```diff
@@ -21,8 +22,12 @@
 
 
 class FunctionalFinderTests(testtools.TestCase):
+    # NOTE(nic): For test runners like PyCharm, set the environment variable
+    #  `RUN_FUNCTIONAL_TESTS=0` in the main test configuration to skip these
+    #  tests.  tox/stestr should already ignore them unless specifically asked.
     def setUp(self):
-        self.skipTest("functional tests disabled for unit test runs")
+        if not int(os.environ.get('RUN_FUNCTIONAL_TESTS', 1)):
+            self.skipTest("functional tests disabled for unit test runs")
```
Comment (juledwar) [inline soufi/tests/functional/test_functional.py:30]: I think a better way of doing this is to move the `functional` dir up a level to `soufi/functional` and have two separate trees of tests. Then you don't need all the hacky environment variable & stestr exclusions.

File: soufi/tests/finders/test_ubuntu_finder.py (hunk 10 above)
Comment (0xDEC0DE) [reply soufi/tests/finders/test_ubuntu_finder.py:39]: The *casus belli* for this entire change is putting caching behavior in the hands of the caller, and the library aggressively not having an opinion about it happening anymore, but OK.  It's easy enough to do.

File: soufi/tests/finders/test_yum_finder.py (hunk 11 above)
Comment (juledwar) [reply soufi/tests/finders/test_yum_finder.py:415]: Much more betterer, thanks.

File: soufi/tests/finders/test_ubuntu_finder.py (hunk 10 above)
Comment (juledwar) [reply soufi/tests/finders/test_ubuntu_finder.py:39]: In this case I don't think it should have moved out of lru_cache, but the fact that there's a unified caching infra now made sense. However, it's a blunt tool as it stands as you either get ALL THE THINGS or nothing. This one should generally always get memory cached as it's a low overhead cache but a very slow call.

File: soufi/tests/functional/test_functional.py (hunk 13 above)
Comment (0xDEC0DE) [reply soufi/tests/functional/test_functional.py:30]: Eww.  But OK.

**Review by juledwar — APPROVED** [review]
